### ISSUE TITLE
Request/Discuss: Weighted Signers

### ISSUE THREAD

**Issue by kamescg** [coinbase/smart-wallet#21]
Today most smart accounts implementations treat every signer as equal, which generally makes sense for multi-sigs where each user is expected to be an unique individual and wants equal signing rights.

But I question that design constraint as it relates smart accounts, which are designed for a single person to use across multiple environments and contexts.

For example the Light smart account implementation by @shunkakinoki includes the ability to add `weights` to each signer.

https://light.so/0xFbd80Fe5cE1ECe895845Fd131bd621e2B6A1345F/owners

![image](https://github.com/coinbase/smart-wallet/assets/3408362/91a4e49c-ff67-49e5-8d72-4ab536825009)

The weights give certain accounts increased/decreased permissions when it comes to authorizing transactions.

For example if the smart account has a threshold of 2/N it would require two accounts with a weight of 1 or a single account with a weight of 2 to authorize a transaction.

I think this starts to get especially interesting especially when you assign "transaction rules" around different signing thresholds and account weights.

For example maybe there is a spending limit of 1,000 USDC every 14 days for accounts with a weight of 1 and thus don't require additional signatures to confirm the transaction, but for any transaction that moving 1,000+ USDC it would require a signature from multiple accounts.

# tl;dr
Having equally weighted signers in a multi-sig is a popular approach for smart accounts today. But it's worth questioning if that design constraint is worth keeping as move into the era of smart account for individuals.

- How can weighted accounts improve the smart account experience?
- Is it worth also exploring "transaction rules" around different signing thresholds?
- Can these features be added as a module, instead of the core smart account logic?

**Comment by wilsoncusack** [comment]
Thanks for this! Our goal with multiple signers was to make it easy for a user to control the same account across devices and in different contexts. I think I would implement a weighted signer at the periphery: have an owner that is itself a contract that has this logic.

**Comment by kamescg** [comment]
That makes sense. Could definitely see that working easily with a periphery account and handling the core logic there.

**Comment by wilsoncusack** [comment]
Going to close this for now, as I am not sure we'll ever take on in the core. Thank you!
